watches.go

```diff
@@ -52,39 +66,70 @@ func (wm *watchManager) handleWatchRequest(w http.ResponseWriter, r *http.Reques
 		http.Error(w, fmt.Sprintf("invalid 'fromIndex': %v", err), http.StatusBadRequest)
 		return
 	}
-
-	go wm.manageWatch(conn, topic, genID, idx)
+	log.Printf("Connection accepted from %v", r.RemoteAddr)
+	if err = wm.manageWatch(w, topic, genID, idx); err != nil {
+		log.Printf("Error: watch %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 }
 
-func (wm *watchManager) manageWatch(conn *websocket.Conn, topic, genID string, idx uint64) {
-	log.Printf("Connection accepted from %v", conn.RemoteAddr())
-	defer closeConn(conn)
+func (wm *watchManager) manageWatch(w http.ResponseWriter, topic string, genID string, idx uint64) error {
+	cn, canNotifyClose := w.(http.CloseNotifier)
+	flusher, canFlush := w.(http.Flusher)
+	if !canNotifyClose || !canFlush {
+		return errors.New("Error: cannot stream")
+	}
+	aw := newActiveWatch(wm, topic, genID, idx, httputil.NewChunkedWriter(w), flusher)
 	for {
-		msgsResponse, err := wm.store.get(topic, genID, idx)
-		if err != nil {
-			handleError(err, conn)
-			return
-		}
-		if msgsLength := len(msgsResponse.Messages); msgsLength > 0 {
-			if err := conn.WriteJSON(msgsResponse); err != nil {
-				handleError(err, conn)
-				return
+		select {
+		case <-cn.CloseNotify():
+			aw.close()
+			return nil
+		default:
+			if err := aw.handleNewMessages(); err != nil {
+				return err
```

Comment:
aw is not closed here. I am not sure if it is important or not. Maybe you can use `defer aw.close()` ?